chia/util/db_wrapper.py

```diff
@@ -259,6 +259,20 @@ async def writer_maybe_transaction(self) -> AsyncIterator[aiosqlite.Connection]:
                 finally:
                     self._current_writer = None
 
+    @contextlib.asynccontextmanager
+    async def reader(self) -> AsyncIterator[aiosqlite.Connection]:
+        async with self.reader_no_transaction() as connection:
+            if connection.in_transaction:
+                yield connection
+            else:
+                await connection.execute("BEGIN DEFERRED;")
+                try:
+                    yield connection
+                finally:
+                    # close the transaction with a rollback instead of commit just in
+                    # case any modifications were submitted through this reader
+                    await connection.rollback()
```

Comment:
similar to my comment above, I believe this will roll back *all* transactions, not just this one.
consider the case where we have one task that has a write-connection (i.e. `SAVEPOINT) open, and writing to it. Within this transaction it opens a `reader` transaction which throws. Just leaving *this* context shouldn't roll back all previous changes within the outer transaction.

I think this only works in the scenario where `reader_no_transaction()` returns one of the reader-connections. But remember that if the calling task currently also has the writer, you get the writer connection back.

I suspect to get this right it might be simpler to not rely on `reader_no_transaction()`